Always inline InterpCx::layout_of after perf regression

Stypox opened the pull request:
Followup to rust-lang/rust#142721 to fix the performance regression. I ran one quick benchmark locally (`ctfe-stress-5`) and it does seem to be faster. I further tried adding `#[inline(always)]` to `compiler/rustc_middle/src/ty/layout.rs` in `layout_of()` under `LayoutOfHelpers` but that didn't change the benchmark results at all.

@rust-timer build e7e3c9e3d687d167197514724546ca99e5e72391

(I'm not sure I have permission to do the above)

r? @RalfJung 

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

Stypox committed:
Always inline InterpCx::layout_of after perf regression

RalfJung commented:
That's not quite the right command -- and indeed you won't have permission.

@bors2 try
@rust-timer-queue

RalfJung commented:
Argh...
@rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

RalfJung commented:
It seems potentially non-trivial to optimize away the drop glue for this `_span` variable for the CTFE instance of the machine. I wonder if it would help to make that analysis easier for the compiler: instead of a `TRACING_ENABLED` associated constant, we could have:
```rust
type ENTERED_TRACING_SPAN = ();
fn enter_tracing_span(tracing::Span) -> Self::ENTERED_TRACING_SPAN;
```
The macro then simply expands to `$machine::enter_tracing_span(tracing::info_span!($($tt)*))`.

RalfJung commented:
@rustbot author

Stypox committed:
Replace TRACING_ENABLED with enter_trace_span()

Hopefully this will make tracing calls be optimized out properly when tracing is disabled

RalfJung commented:
> I couldn't use this directly as "associated type defaults are unstable" and, most importantly, since trait implementations would be able to override ENTERED_TRACING_SPAN without overriding enter_tracing_span, () would not always have the Self::ENTERED_TRACING_SPAN type. 

Ah, fair.

> Anyway, I used this instead:

Interesting, I would have just removed the default. ;)

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued ce51ffbeae12a0e63028938a74cbb1da8c986ffe with parent f0b67dd97d74610ee4185cf01c775a563c2017a2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f0b67dd97d74610ee4185cf01c775a563c2017a2&end=ce51ffbeae12a0e63028938a74cbb1da8c986ffe).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.1 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (ce51ffbeae12a0e63028938a74cbb1da8c986ffe): [comparison URL](https://perf.rust-lang.org/compare.html?start=f0b67dd97d74610ee4185cf01c775a563c2017a2&end=ce51ffbeae12a0e63028938a74cbb1da8c986ffe&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=f0b67dd97d74610ee4185cf01c775a563c2017a2&end=ce51ffbeae12a0e63028938a74cbb1da8c986ffe&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.1%, 0.8%]   | 15    |
| Improvements ✅ <br /> (primary)   | -2.9% | [-2.9%, -2.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-1.0%, -0.5%] | 3     |
| All ❌✅ (primary)                 | -2.9% | [-2.9%, -2.9%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=f0b67dd97d74610ee4185cf01c775a563c2017a2&end=ce51ffbeae12a0e63028938a74cbb1da8c986ffe&stat=max-rss)
<details>
<summary>Results (primary -3.9%, secondary -0.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.5%  | [3.5%, 3.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -3.9% | [-3.9%, -3.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -4.8% | [-4.8%, -4.8%] | 1     |
| All ❌✅ (primary)                 | -3.9% | [-3.9%, -3.9%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=f0b67dd97d74610ee4185cf01c775a563c2017a2&end=ce51ffbeae12a0e63028938a74cbb1da8c986ffe&stat=cycles:u)
<details>
<summary>Results (secondary -1.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.6%  | [2.3%, 4.8%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -7.3% | [-7.5%, -7.0%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=f0b67dd97d74610ee4185cf01c775a563c2017a2&end=ce51ffbeae12a0e63028938a74cbb1da8c986ffe&stat=size:linked_artifact)
<details>
<summary>Results (primary -1.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.1% | [-1.1%, -1.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.1% | [-1.1%, -1.1%] | 1     |
</details>

**Bootstrap**: 460.935s -> 461.093s (0.03%)
**Artifact size**: 372.19 MiB -> 371.92 MiB (-0.07%)

RalfJung commented:
Fun, so that does seem to help for ctfe-stress, but other benchmarks don't like it, and it doesn't fully negate the effect of https://github.com/rust-lang/rust/pull/143334. Can you try adding `inline(always)` to the machine hook as well?

The only other thing I can think of is an entirely different, closure-based API. But given that the regression only affects secondary benchmarks / stress tests, I don't think that's worth it.

Stypox committed:
Add inline(always) to Machine::enter_trace_span

Stypox commented:
I added inline(always) to the machine hook and also opened https://github.com/rust-lang/rust/pull/143520 as an alternative using a closure (I'm not sure it's what you had in mind but given the simple change I wanted to try it anyway). Could you run the benchmarks on both? Thanks!

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued e5441bddc082141fd7bdd42b5e72766016324a9c with parent febb10d0a2d29278135676783f6a22eb83295981, future [comparison URL](https://perf.rust-lang.org/compare.html?start=febb10d0a2d29278135676783f6a22eb83295981&end=e5441bddc082141fd7bdd42b5e72766016324a9c).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e5441bddc082141fd7bdd42b5e72766016324a9c): [comparison URL](https://perf.rust-lang.org/compare.html?start=febb10d0a2d29278135676783f6a22eb83295981&end=e5441bddc082141fd7bdd42b5e72766016324a9c&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=febb10d0a2d29278135676783f6a22eb83295981&end=e5441bddc082141fd7bdd42b5e72766016324a9c&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.1%, 0.8%]   | 16    |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.6% | [-1.0%, -0.3%] | 4     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=febb10d0a2d29278135676783f6a22eb83295981&end=e5441bddc082141fd7bdd42b5e72766016324a9c&stat=max-rss)
<details>
<summary>Results (primary -1.6%, secondary -2.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-1.6%, -1.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -1.6% | [-1.6%, -1.6%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=febb10d0a2d29278135676783f6a22eb83295981&end=e5441bddc082141fd7bdd42b5e72766016324a9c&stat=cycles:u)
<details>
<summary>Results (primary -1.2%, secondary 3.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.1%  | [3.1%, 3.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.2%, -1.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.2% | [-1.2%, -1.2%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=febb10d0a2d29278135676783f6a22eb83295981&end=e5441bddc082141fd7bdd42b5e72766016324a9c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 460.756s -> 461.178s (0.09%)
**Artifact size**: 372.14 MiB -> 371.86 MiB (-0.08%)

RalfJung commented:
I agree, let's close in favor of https://github.com/rust-lang/rust/pull/143520.
